src/engine/MaxConstraint.cpp

```diff
@@ -121,64 +123,83 @@ void MaxConstraint::registerAsWatcher( ITableau *tableau )
 void MaxConstraint::unregisterAsWatcher( ITableau *tableau )
 {
     for ( unsigned element : _initialElements )
+    {
         tableau->unregisterToWatchVariable( this, element );
+        if ( _elementToAux.exists( element ) )
+            tableau->unregisterToWatchVariable( this, _elementToAux[element] );
+    }
 
     if ( !_initialElements.exists( _f ) )
         tableau->unregisterToWatchVariable( this, _f );
 }
 
 void MaxConstraint::notifyVariableValue( unsigned variable, double value )
 {
-    if ( ( _elements.exists( _f ) || variable != _f ) &&
-         ( !maxIndexSet() || !_assignment.exists( getMaxIndex() ) || _assignment.get( getMaxIndex() ) < value ) )
-    {
-        setMaxIndex( variable );
-    }
     _assignment[variable] = value;
 }
 
 void MaxConstraint::notifyLowerBound( unsigned variable, double value )
 {
     if ( _statistics )
-        _statistics->incLongAttribute( Statistics::NUM_BOUND_NOTIFICATIONS_TO_PL_CONSTRAINTS );
+        _statistics->incLongAttribute
+            ( Statistics::NUM_BOUND_NOTIFICATIONS_TO_PL_CONSTRAINTS );
 
-    if ( existsLowerBound( variable ) && !FloatUtils::gt( value, getLowerBound( variable ) ) )
+    if ( existsLowerBound( variable ) &&
+         !FloatUtils::gt( value, getLowerBound( variable ) ) )
         return;
 
     setLowerBound( variable, value );
 
-    bool maxErased = false;
-
-    if ( _elements.exists( variable ) && FloatUtils::gt( value, _maxLowerBound ) )
+    /*
+      See if we can eliminate any cases.
+    */
+    if ( _auxToElement.exists( variable ) )
     {
-        _maxLowerBound = value;
-        List<unsigned> toRemove;
-        for ( auto element : _elements )
+        // The case that this variable is an aux variable.
+        // We can eliminate the corresponding case if the aux variable is
+        // positive.
+        if ( FloatUtils::isPositive( value ) )
```

Comment:
The ifs on line 156 and 162 can merged. Comment on line 158 seems redundant.

Re: Comment style - we should probably use /*  */  for function descriptions (here used on line 153-155), and use // for line comments in the actual implementation. The extra space used by /**/ is useful in header for the reader to distinguish (groups of) methods visually, while the same amount of space within code itself would be detracting from easy readability.

Reply:
Done.